Tasks/JavaToolInstallerV0/javatoolinstaller.ts

```diff
@@ -81,35 +85,133 @@ async function getJava(versionSpec: string) {
+    const newVolumes: string[] = fs.readdirSync(VOLUMES_FOLDER).filter(volume => !volumes.has(volume));
 
-    return extractSource;
+    if (newVolumes.length !== 1) {
+        throw new Error(taskLib.loc('UnsupportedDMGStructure'));
+    }
+    return path.join(VOLUMES_FOLDER, newVolumes[0]);
 }
 
-function getFileEnding(file: string): string {
-    let fileEnding = '';
-
-    if (file.endsWith('.tar')) {
-        fileEnding = '.tar';
-    } else if (file.endsWith('.tar.gz')) {
-        fileEnding = '.tar.gz';
-    } else if (file.endsWith('.zip')) {
-        fileEnding = '.zip';
-    } else if (file.endsWith('.7z')) {
-        fileEnding = '.7z';
+/**
+ * Get path to a .pkg file.
+ * Only for macOS.
+ * @param volumePath Path to the folder containing a .pkg file.
+ */
+function getPackagePath(volumePath: string): string {
+    const packages: string[] = fs.readdirSync(volumePath).filter(file => file.endsWith('.pkg'));
+
+    if (packages.length === 1) {
+        return path.join(volumePath, packages[0]);
+    } else if (packages.length === 0) {
+        throw new Error(taskLib.loc('NoPKGFile'));
     } else {
-        throw new Error(taskLib.loc('UnsupportedFileExtension'));
+        throw new Error(taskLib.loc('SeveralPKGFiles'));
     }
+}
+
+async function installPkg(pkgPath: string, extendedJavaHome: string, versionSpec: string): Promise<string> {
+    if (!fs.existsSync(pkgPath)) {
+        throw new Error('PkgPathDoesNotExist');
+    }
+
+    console.log(taskLib.loc('InstallJDK'));
 
-    return fileEnding;
+    // Using set because 'includes' array method requires tsconfig option "lib": ["ES2017"]
+    const JDKs: Set<string> = new Set(fs.readdirSync(JDK_FOLDER));
+
+    await runPkgInstaller(pkgPath);
+
+    const newJDKs = fs.readdirSync(JDK_FOLDER).filter(jdkName => !JDKs.has(jdkName));
+
+    let jdkDirectory: string;
+
+    if (newJDKs.length === 0) {
+        const preInstalledJavaDirectory: string | undefined = taskLib.getVariable(extendedJavaHome);
+        if (!preInstalledJavaDirectory) {
+            throw new Error(taskLib.loc('JavaNotPreinstalled', versionSpec));
+        }
+        console.log(taskLib.loc('PreInstalledJavaUpgraded'));
+        console.log(taskLib.loc('UsePreinstalledJava', preInstalledJavaDirectory));
+        jdkDirectory = preInstalledJavaDirectory;
+    } else {
+        console.log(taskLib.loc('JavaSuccessfullyInstalled'));
+        jdkDirectory = path.join(JDK_FOLDER, newJDKs[0], JDK_HOME_FOLDER);
+    }
+
+    return jdkDirectory;
+}
+
+/**
+ * Install a .pkg file.
+ * Only for macOS.
+ * @param pkgPath Path to a .pkg file.
+ */
+async function runPkgInstaller(pkgPath: string): Promise<void> {
+    try {
+        const installer = sudo('installer');
+        installer.line(`-package "${pkgPath}" -target /`);
+        await installer.exec();
+    } catch (error) {
+        taskLib.debug('Failed to install pkg file');
+        taskLib.error(error.message);
```

Comment:
Doen't it only prints an error, but continue execution?

Reply:
Removed try-catch to catch the error on the top level. Thanks!